### DISCUSSION TITLE
Plotting multiple individual tubes on a single plot

### DISCUSSION THREAD

**Discussion by jpmorr** [pyvista/pyvista#1678] [Q&A] [answered]
I'm trying to create a plot of a network of lines between points where each line has it's own weighted thickness and scalar value. I've put together a solution (below) that works, albeit very slowly. It does what I want but it's not really suitable for plotting the number tubes I need to handle. It's currently taking several minutes to plot a few thousand tubes and I typically want to do this for 10's or 100's of thoudands of tubes at a time.

Here's an example of the network I'm trying to plot, where the lines are coloured and thickness weighted by some query value:
![image](https://user-images.githubusercontent.com/10863809/134036808-0e462a34-254f-4a72-aa44-7009e8df28ee.png)

Here's what I'm currently doing - I've got two functions, one that creates some poly data from a `numpy` array of points (start and end point of line) and another that loops though arrays of start and end points for each line and assigns a value to the query.

````
def lines_from_points(points):
    """Given an array of points, make a line set"""
    poly = pv.PolyData()
    poly.points = points
    cells = np.full((len(points) - 1, 3), 2, dtype=np.int_)
    cells[:, 1] = np.arange(0, len(points) - 1, dtype=np.int_)
    cells[:, 2] = np.arange(1, len(points), dtype=np.int_)
    poly.lines = cells
    return poly


def create_tubes(start_point, end_point, query, streamline_width=0.0001):
    tubes = []

    norm_query = query / np.max(np.abs(query))

    # Create tubes
    for element in range(len(query)):
        points = np.vstack([start_point[element], end_point[element]])
        line = lines_from_points(points)
        line["query"] = np.array([query[element]])

        tubes.append(line.tube(radius=streamline_width * norm_query[element]))

    return tubes

````

Once I've created a list of all the indivdual tubes I have to loop through each one and add it to the plot:
````
tubes = create_tubes(start_point, end_point, query_data , streamline_width=0.0001)

# Change plotting theme
pv.set_plot_theme("document")  ## white background

# create Plot
plotter = BackgroundPlotter()
# plotter = pv.Plotter(notebook=False)

for tube in tubes:
    plotter.add_mesh(tube, scalars="query", cmap="inferno")
plotter.show_axes()
````
This is the part that is extremly slow becuase it's a plotting loop and it would be great if there was a way to plot all of the tubes at once rather than adding each one as a new mesh. I don't think I can plot all as a single spline as there may be many unconnected groups in the full datasets of many thousand points.

Here's the first 25 points in the network in the image as sample data:
````
start_point = np.array([[-0.01282559, -0.01287017,  0.00268859],
       [-0.01282559, -0.01287017,  0.00268859],
       [-0.01282559, -0.01287017,  0.00268859],
       [-0.01282559, -0.01287017,  0.00268859],
       [-0.01282559, -0.01287017,  0.00268859],
       [-0.01340175, -0.01342823,  0.00451884],
       [-0.01340175, -0.01342823,  0.00451884],
       [-0.00798815, -0.0171951 ,  0.00233246],
       [-0.00798815, -0.0171951 ,  0.00233246],
       [-0.00798815, -0.0171951 ,  0.00233246],
       [-0.00557685, -0.0181246 ,  0.00267036],
       [-0.01124414, -0.01530663,  0.00099935],
       [-0.01124414, -0.01530663,  0.00099935],
       [-0.01124414, -0.01530663,  0.00099935],
       [-0.01199464, -0.01468645,  0.00274531],
       [-0.01199464, -0.01468645,  0.00274531],
       [-0.01199464, -0.01468645,  0.00274531],
       [-0.01096209, -0.0133272 ,  0.00099998],
       [-0.01096209, -0.0133272 ,  0.00099998],
       [-0.00960519, -0.01426426,  0.00213129],
       [-0.00822006, -0.01515808,  0.00099916],
       [-0.00822006, -0.01515808,  0.00099916],
       [-0.00614223, -0.01625568,  0.00219329],
       [-0.00762747, -0.01349969,  0.00195301],
       [-0.00762747, -0.01349969,  0.00195301]])

end_point = np.array([[-0.01340175, -0.01342823,  0.00451884],
       [-0.01237146, -0.011909  ,  0.00099722],
       [-0.01215237, -0.01187416,  0.00428584],
       [-0.01199464, -0.01468645,  0.00274531],
       [-0.012751  , -0.01087287,  0.00266228],
       [-0.014442  , -0.01231921,  0.00581654],
       [-0.01196508, -0.01309501,  0.00586829],
       [-0.00888431, -0.01677514,  0.00406928],
       [-0.00662112, -0.01778563,  0.00099905],
       [-0.00967911, -0.01631412,  0.00172991],
       [-0.00662112, -0.01778563,  0.00099905],
       [-0.01096209, -0.0133272 ,  0.00099998],
       [-0.00967911, -0.01631412,  0.00172991],
       [-0.01199464, -0.01468645,  0.00274531],
       [-0.01068517, -0.01347768,  0.00365126],
       [-0.01190346, -0.01476117,  0.00474095],
       [-0.01034458, -0.0155447 ,  0.00347795],
       [-0.01237146, -0.011909  ,  0.00099722],
       [-0.00960519, -0.01426426,  0.00213129],
       [-0.01034458, -0.0155447 ,  0.00347795],
       [-0.00960519, -0.01426426,  0.00213129],
       [-0.00967911, -0.01631412,  0.00172991],
       [-0.00662112, -0.01778563,  0.00099905],
       [-0.00623212, -0.01444268,  0.00303074],
       [-0.00911439, -0.01256266,  0.00099974]])

query_data = np.array([1.70297489e-06, 2.29733079e-06, 8.84168147e-07, 1.86474461e-06,
       1.13034644e-06, 9.83076761e-07, 9.86501195e-07, 9.97705005e-07,
       1.12835375e-06, 3.58541341e-07, 3.32862219e-07, 5.77620311e-07,
       4.83357361e-07, 9.36432225e-07, 8.57709316e-07, 8.86824668e-07,
       9.87560144e-07, 5.98387852e-07, 2.19303789e-07, 6.95079425e-08,
       1.98071403e-07, 1.86493765e-07, 9.27577233e-07, 5.65600421e-07,
       5.77192380e-07])


````

I think there must be a better (quicker) way of doing this and achieving a similar result? 

I'm also wondering wheter it's possible to add a different opcaity to each tube dependent on the weighting value?

**Comment by MatthewFlamm** [comment]
Glyph mapping is much faster than individually plotting each tube.  You would use a cylinder and scale each dimension according to length and diameter.  It seems possible to do this in vtk according to [this example](https://kitware.github.io/vtk-examples/site/Cxx/Visualization/Glyph3DMapper/).  This example uses the method `SetScaleModeToScaleByVectorComponents`, which is unused in PyVista, but could be added as a feature for a future version.

**Reply by MatthewFlamm** [reply]
See https://docs.pyvista.org/api/core/_autosummary/pyvista.DataSetFilters.glyph.html?highlight=glyph#pyvista.DataSetFilters.glyph, but as I mentioned when you pass in a 3D value as the `scale`, I believe PyVista will currently take the magnitude of that 3D value instead of each component separately.

**Reply by jpmorr** [reply]
Plotting cylinder glyphs is possibly the way to do this quickly, and I need to spend some time trying out that and seeing how best to control the length and radius. But if memory serves me correctly, the glyphs size changes as you zoom in and out on the image?

**Comment by darikg** [comment]
@MatthewFlamm's suggestion seems like the way to go in the long run, but as a kludge is this any faster?

```python
mesh = PolyData()
for tube in create_tubes(...):
    mesh = mesh.merge(tube)   # `.merge` is `.boolean_add` in pyvista < 0.32

plotter.add_mesh(mesh)
````

**Reply by akaszynski** [reply]
It's better to create all the tubes first and then merge later:

```py
tubes = [tube for tube in create_tubes(...)]
mesh = tube[0].merge(tube[1:])  # see note regarding merge vs boolean add
```

**Reply by jpmorr** [reply]
`merge` makes things very quick and creating the tubes is now the bottleneck, but as that is only a few seconds, it's good enough for now.

**Reply by akaszynski** [reply]
Can we get our hands on this data or part of it?  This would be a great example for our documentation?

**Reply by jpmorr** [reply]
@akaszynski The data is just a series of points and connecting vectors - really all just simple numpy arrays. I guess I could pickle the arrays and share that? It's actually not too disimilar to https://github.com/pyvista/pyvista-support/issues/494, except that I have a lot more elements in a different format and want to size the elements according to the force in the element.

**Reply by akaszynski** [reply]
I'd just save the numpy arrays and then upload it here as a zip, along with a simple script that shows how you're using them.  I wouldn't send pickled data through the web as it's unsafe.  See: https://www.synopsys.com/blogs/software-security/python-pickling/

**Reply by jpmorr** [reply]
@akaszynski Here's the `numpy` data file and plotting script I'm using at the moment for testing this: 
[Tests.zip](https://github.com/pyvista/pyvista/files/7216315/Tests.zip)

This is a smaller dataset than I normally work on because I'm just testing things. Normally it would be 100,000 or more connections. Also, merge works well here because everything forms just one connected group - sometimes there maybe be multiple unconnected clusters and I'm not sure how merge will work in that case.

**Comment by jpmorr** [comment]
I have a question now that relates to whether `merge` or `MultiBlock` is more suited as I came across this discussion (https://github.com/pyvista/pyvista/issues/1295 ) which shows multiblock being used. Are `merge` and `MultiBlock` nearly the same thing?

**Reply by MatthewFlamm** [reply]
> Are `merge` and `MultiBlock` nearly the same thing?

No. `merge` creates a single merged mesh that in some instances can change the structure of the original meshes.   `MultiBlock` creates a collection of meshes that retain their individual identity, but are grouped together in one container.

**Comment by jpmorr** [comment, marked as answer]
@MatthewFlamm's solution is indeed the quickest method (almost instantaneous), but I'm not able to figure out how to control both the length and radius of the cylinder. If I understand his answer correctly, it's not possible to scale both the length and radius separately in `pyvista`.

```` python
mesh = pv.PolyData(centre_points)
mesh["query"] = query
mesh["vectors"] = orient_vect
mesh["length"] = np.linalg.norm(end_point - start_point, axis=1)

# Low resolution geometry
geom = pv.Cylinder(resolution=20, radius=0.05)

# Perform the glyph
glyphs = mesh.glyph(orient="vectors", scale='length', factor=1, geom=geom)

pl = pv.Plotter()
# pl = BackgroundPlotter()
pl.add_mesh(glyphs, scalars="query", lighting=True)

pl.show_axes()
pl.show()
````
![image](https://user-images.githubusercontent.com/10863809/134175049-ec48d04e-ef5d-4f45-aa06-830e756219d1.png)

Using the `tubes` method with the `merge` solution suggested by @darikg and @akaszynski  is much faster than my original attempt but still many times slower than the `glyph`s. However, it does allow the thickness of the tubes to be set individually, giving the desired result.

```` python
tubes = create_tubes(start_point, end_point, query, streamline_width=0.0005)
mesh = tubes[0].merge(tubes[1:])  # see note regarding merge vs boolean add

# Change plotting theme
pv.set_plot_theme("document")  ## white background

# create Plot
plotter = BackgroundPlotter()
plotter.add_mesh(mesh, scalars="query", cmap="viridis")
plotter.show_axes()

````
![image](https://user-images.githubusercontent.com/10863809/134177128-08f9ec65-2104-447b-97b4-bf38a065116f.png)

One other thing I have noticed is that the `tubes` solution seems to cause the plot window to freeze and not respond quite a lot, possibly becuase of the number of points created for each tube - about 80 points for a straight tube generated from a line with just 2 points. The `tubes` solution had about 863,960 points against just 21,600 for the `glyph`s based solution.

So is it possible to achieve the varying length and thickness with glyphs? If so, that would be the ideal solution. Otherwise it's a compromise between viually getting the desired result slowly and maybe struggle to interract with becuase of the number of points, or use the glyphs to have uniform radius and lose some information.

Hope there is a solution!

**Reply by MatthewFlamm** [reply]
I think the best solution is to add the functionality of using `SetScaleModeToScaleByVectorComponents` to the pyvista glyph filter.  I think it would be appropriate to add an optional keyword argument for this purpose.

**Reply by jpmorr** [reply]
@MatthewFlamm That would make things much easier. Do I need to make this an issue?

**Reply by MatthewFlamm** [reply]
If you are interested, you can also submit a PR to do so.  If not, I'm also interested in this functionality for more control over arrow glyph plots.

**Reply by jpmorr** [reply]
@MatthewFlamm I don't quite have the expertise to make the required change I think, even if it is relatively simple. I'll submit an issue / feature request.

**Comment by MatthewFlamm** [comment]
So I tested my suggestion, I don't think it will work unfortunately.  My only suggestion now is to break your set of lines into  bins of scaling sizes and use glyphs to generate a set of meshes to plot. In outline format:

This way you get the speed of glyph generation and plotting, but also the flexibility to scale the size of the glyph.  More bins -> more realistic rendering, less bins -> faster rendering.

```
Select 10 bins of scaling
For each bin
    Find the set of lines according to this scaling size
    Create Glyphs with a single size and correct orientation
    Add glyphs to plotter
```

**Reply by jpmorr** [reply]
Could you do this using the `MultiBlock` or `merge` method that worked for the tubes? I'm not sure it would make it any faster though? 

I think I will try something like you have suggested but I've kind of settled on the idea of using glyphs as much as possible and only using the tube rendering when it's really important to have the variation in tube diameter - in some cases there is little difference in the diameters so an intelligent initial check can be used to decide if I need to plot with the tubes.

**Reply by jpmorr** [reply]
@MatthewFlamm Actually, I just saw this discussion (https://github.com/pyvista/pyvista-support/issues/494) about plotting members in a truss, which is a simple example where all the members have the same width - would this method accept an array of different widths so that each one could have a different width? That would almost be the ideal solution.

**Reply by JacobBumgarner** [reply]
@jpmorr Did you make any other observations about this process? I've been working with plotting datasets that are extremely similar to yours. I've been using MultiBlocks but still wish there was another faster option.

**Reply by jpmorr** [reply]
@JacobBumgarner I've stuck to using the glyphing method I posted above for most of my plotting as the majority of the time the monosized cylinders are good enough to tell what's happening. I can then use the tubes based method if I really want to have them sized by a scalar. It's not the perfect solution but it works. It would be ideal if the truss example could also just take an array of sizes for each cylinder, but I think it might not be possible because vtk doesn't support this.
